advanced_source/cpp_autograd.rst

```diff
@@ -0,0 +1,442 @@
+Let's see an example of it using ``torch::autograd::grad``:
+
+.. code-block:: cpp
+
+  #include <torch/torch.h>
+  
+  auto model = torch::nn::Linear(4, 3);
+  
+  auto input = torch::randn({3, 4}).requires_grad_(true);
+  auto output = model(input);
+  
+  // Calculate loss
+  auto target = torch::randn({3, 3});
+  auto loss = torch::nn::MSELoss()(output, target);
+  
+  // Use norm of gradients as penalty
+  auto grad_output = torch::ones_like(output);
+  auto gradient = torch::autograd::grad({output}, {input}, /*grad_outputs=*/{grad_output},
+                  /*retain_graph=*/true, /*create_graph=*/true,
+                  /*allow_unused=*/true)[0];
+  gradient = gradient.view({-1, 1});
+  auto gradient_penalty = torch::pow((gradient.norm(2, /*dim=*/1) - 1), 2).mean();
+  
+  // Add gradient penalty to loss
+  auto combined_loss = loss + gradient_penalty;
+  combined_loss.backward();
+  
+  std::cout << input.grad() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  -0.2574  0.0494  0.3060  0.3215
+  -0.1197  0.0248  0.1405  0.1401
+  [ CPUFloatType{2,4} ]
+
+Please see the documentation for ``torch::autograd::backward``
+(`link <https://pytorch.org/cppdocs/api/function_namespacetorch_1_1autograd_1afa9b5d4329085df4b6b3d4b4be48914b.html>`_)
+and ``torch::autograd::grad``
+(`link <https://pytorch.org/cppdocs/api/function_namespacetorch_1_1autograd_1a1e03c42b14b40c306f9eb947ef842d9c.html>`_)
+for more information on how to use them.
+
+Using custom autograd function in C++
+-------------------------------------
+
+(Adapted from `this tutorial <https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd>`_)
+
+Adding operations to ``torch::autograd`` requires implementing a new ``torch::autograd::Function``
+subclass for each operation. ``torch::autograd::Function`` s are what ``torch::autograd``
+uses to compute the results and gradients, and encode the operation history. Every
+new function requires you to implement 2 methods: ``forward`` and ``backward``, and
+please see `this link <https://pytorch.org/cppdocs/api/structtorch_1_1autograd_1_1_function.html>`_
+for the detailed requirements.
+
+Below you can find code for a ``Linear`` function from ``torch::nn``:
+
+.. code-block:: cpp
+
+  #include <torch/torch.h>
+  
+  using namespace torch::autograd;
+  
+  // Inherit from Function
+  class LinearFunction : public Function<LinearFunction> {
+   public:
+    // Note that both forward and backward are static functions
+  
+    // bias is an optional argument
+    static Variable forward(AutogradContext *ctx, Variable input, Variable weight, Variable bias = Variable()) {
+      ctx->save_for_backward({input, weight, bias});
+      auto output = input.mm(weight.t());
+      if (bias.defined()) {
+        output += bias.unsqueeze(0).expand_as(output);
+      }
+      return output;
+    }
+  
+    static variable_list backward(AutogradContext *ctx, variable_list grad_outputs) {
+      auto saved = ctx->get_saved_variables();
```

Comment:
Why is this not `get_saved_tensors()` ?

Reply:
Yeah I think this should be `get_saved_tensors()`, and my guess is that this API was created before the Variable/Tensor merge was completed, and the author wanted to make a distinction between tensors and variables. Now that we don't have this distinction anymore, I think we should change it to match the Python API: https://github.com/pytorch/pytorch/issues/36099.

Reply:
I renamed all `Variable`s to `Tensor` now, to minimize the inconsistency with Python to only this `get_saved_variables()` function :)

Reply:
Yes I guess it is just a remain of the cpp Tensor/Variable merge.